cranelift/interpreter/src/interpreter.rs

```diff
@@ -310,23 +371,62 @@ impl<'a> State<'a, DataValue> for InterpreterState<'a> {
         Address::from_parts(size, AddressRegion::Stack, 0, final_offset)
     }
 
-    fn heap_address(&self, _size: AddressSize, _offset: u64) -> Result<Address, MemoryError> {
-        unimplemented!()
+    fn heap_address(
+        &self,
+        size: AddressSize,
+        heap: Heap,
+        offset: u64,
+    ) -> Result<Address, MemoryError> {
+        // It is possible for this function to be called before the interpreter has started
+        // execution (see `InterpreterState::get_heap_address`). If this happens we cannot
+        // perform global value resolution, and have to build an address directly referencing
+        // the `heap` that was provided.
+        let has_current_function = self.frame_stack.len() == 0;
+        let addr = if has_current_function {
+            let heap_id = heap.as_u32() as u64;
+            Address::from_parts(size, AddressRegion::Heap, heap_id, offset)?
```

Comment:
I'm not entirely sure this belongs here.

I had this separately in `InterpreterState::get_heap_address`, but I didn't want to duplicate the bounds checking done on this function.

Since bounds checking is only required as part of the runtime guarantees (for `heap_address` in addressable memory), I think it would make sense to put this back in `get_heap_address`, but also means we then have `InterpreterState::get_heap_address` and `InterpreterState::heap_address` which do something different.